### PR TITLE
Feature to rotate files periodically after a specified number of windows have elapsed

### PR DESCRIPTION
**Pull request by PramodSSImmaneni**
A feature to rotate files periodically. The rotation period is specified as a parameter.

### CODE REVIEW

**Comment by chandnisingh** [comment]
Pramod,
After looking at the code changes I feel this makes the base implementation complicated. We now have 2 ways of file rotation in the same base implementation.
The existing one on file size is  a most common one and the one you added maybe needed by some.
I think if we could make file rotation a pluggable component with default one based on size then that will allow users to plug in any strategy they want.
We can then also add the strategy based on app window in the library.

**Comment by PramodSSImmaneni** [comment]
Both may not be exclusive, for example both length base and time based rotation may be needed together. Also rotations may be triggered at different points for different strategies for example length based rotation can happen anytime such as whenever tuples are recieved and time based would happen at window. These factors means it may not be a simple plugin strategy and would require a more advanced implementation. Both length based and time based rotations are basic rotation strategies that people think of first when they think of rotation in my opinion and am in favor of providing them in the operator now and look at implementing a comprehensive plugin mechanism in the future.

**Comment by tweise** [comment]
+1 Combination of time and size based rolling is quite common.

File: library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java [hunk 1]
```diff
@@ -185,6 +200,11 @@
   protected StreamCodec<INPUT> streamCodec;
 
   /**
+   * Number of windows since the last rotation 
+   */
+  private int rotationCount;
+
+  /**
```
Comment (chandnisingh) [inline library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java]: Can this not be different for different files? 
So if file x gets open in window 1 and file y gets open in window 5 will  we rotate them both in window 10 (rotationWindows=10)

File: library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java [hunk 2]
```diff
@@ -656,6 +700,39 @@ public void endWindow()
...
+        rotationCount = 0;
+        // Rotate the files
+        Iterator<String> iterator = streamsCache.asMap().keySet().iterator();
+        while (iterator.hasNext()) {
+          String filename = iterator.next();
+          // Rotate the file if the following conditions are met
+          // 1. The file is not already rotated during this period for other reasons such as max length is reached
+          //     or rotate was explicitly called externally
+          // 2. The file is not empty
+          RotationState rotationState = (RotationState)rotationStates.get(filename);
```
Comment (chandnisingh) [inline library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java]: Sorry about nitpicking but casting to RotationState is redundant here.

Comment (PramodSSImmaneni) [inline library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileOutputOperator.java]: My bad I missed the warning. Nitpicking is good :)

**Pull request merged**